File: 528. Random Pick with Weight/pickIndex.js
/**
 * @param {number[]} w
 */
class Solution { 
  constructor(w) {
    this.prefixSum = [];
    let currSum = 0;
    for (let i = 0; i < w.length; i++) {
      currSum += w[i];
      this.prefixSum.push(currSum);
    }

    this.totalSum = currSum;
  };

  pickIndex() {
    const target = this.totalSum * Math.random();
    let left = 0;
    let right = this.prefixSum.length - 1;
    while (left < right) {
      const mid = Math.floor((left + right) / 2);
      if (target > this.prefixSum[mid]) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }

    return left;
  }
}

const w = [1, 3];
const pickupWeight = new Solution(w);
pickupWeight.pickIndex()
/** 
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(w)
 * var param_1 = obj.pickIndex()
 */

File: 528. Random Pick with Weight/pickIndex.py
import random
from typing import List

class Solution:
  def __init__(self, w: List[int]):
    self.prefixSum = []
    currSum = 0
    for weight in w:
      currSum += weight
      self.prefixSum.append(currSum)
    self.totalSum = currSum

  def pickIndex(self) -> int:
    target = self.totalSum * random.random()
    left, right = 0, len(self.prefixSum) - 1
    while left < right:
      mid = left + (right - left) // 2
      if target > self.prefixSum[mid]:
        left = mid + 1
      else:
        right = mid
    
    return left

# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()

w = [1, 3]
obj = Solution(w)
obj.pickIndex()
